### PR TITLE
Add GCS Writer Sink and fix Gcs naming

### PR DESCRIPTION
**Pull request by majetideepak**
Rename GCS to Gcs in code.
Refactor GCS code.
Refactor InsertTests so that both S3 and GCS can reuse the implementation.

### CODE REVIEW

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp [hunk 1]
```diff
@@ -16,170 +16,30 @@
...
-std::string GCSFileSystemTest::object_name_;
-
 TEST_F(GCSFileSystemTest, readFile) {
-  const std::string gcsFile =
-      gcsURI(preexistingBucketName(), preexistingObjectName());
+  const auto gcsFile = gcsURI(
+      testbench_->preexistingBucketName(), testbench_->preexistingObjectName());
 
-  filesystems::GCSFileSystem gcfs(testGcsOptions());
+  filesystems::GCSFileSystem gcfs(testbench_->hiveConfig());
```
Comment (pedroerp) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp:42]: same here. Acronyms are not all upper cases (`GcsFileSystem`)

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+namespace bp = boost::process;
+namespace gc = google::cloud;
+namespace gcs = google::cloud::storage;
+
+namespace facebook::velox::filesystems {
+
+class GCSTestbench : public testing::Environment {
+ public:
+  GCSTestbench() {
+    auto port = std::to_string(facebook::velox::exec::test::getFreePorts(1)[0]);
```
Comment (pedroerp) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h:37]: `getFreePort()` instead

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+          "--port",
+          port,
+          group_);
+      if (serverProcess_.valid() && serverProcess_.running())
+        break;
+      error += " (failed to start)";
+      serverProcess_.terminate();
+      serverProcess_.wait();
+    }
+    if (serverProcess_.valid() && serverProcess_.valid())
```
Comment (pedroerp) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h:70]: is this intended (checking the same condition twice?). Also, curly braces missing

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+#include "velox/connectors/hive/storage_adapters/gcs/GCSUtil.h"
+#include "velox/exec/tests/utils/PortUtil.h"
+
+namespace bp = boost::process;
+namespace gc = google::cloud;
+namespace gcs = google::cloud::storage;
+
+namespace facebook::velox::filesystems {
+
+class GCSTestbench : public testing::Environment {
```
Comment (pedroerp) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h:34]: let's rename this to something more descriptive and intuitive. "bench" implies benchmarking which is not the intent.

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    bucketName_ = "test1-gcs";
+    google::cloud::StatusOr<gcs::BucketMetadata> bucket =
+        client.CreateBucketForProject(
+            bucketName_, "ignored-by-testbench", gcs::BucketMetadata{});
+    ASSERT_TRUE(bucket.ok()) << "Failed to create bucket <" << bucketName_
+                             << ">, status=" << bucket.status();
+
+    objectName_ = "test-object-name";
+    google::cloud::StatusOr<gcs::ObjectMetadata> object =
+        client.InsertObject(bucketName_, objectName_, kLoremIpsum);
```
Comment (pedroerp) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h:126]: should probably move the loremIpsum stuff here.

**Review by pedroerp — COMMENTED** [review]
Thanks @majetideepak. Looks good overall; added a bunch of minor comments.

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp [hunk 6]
```diff
@@ -16,170 +16,30 @@
...
-
-  std::string preexistingObjectName() {
-    return object_name_;
-  }
-
-  std::string preexistingObjectPath() {
-    return preexistingBucketPath() + preexistingObjectName();
+ public:
+  void SetUp() {
+    testbench_ = std::make_shared<GCSTestbench>();
```
[the remark this answers is not in the document]
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp:31]: This uses the storage-testbench library. That is probably why. I renamed it to GcsEmulator
https://github.com/googleapis/storage-testbench

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h (hunk 4 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/gcs/tests/GCSTestbench.h:34]: Named it as `GcsEmulator`.

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp (hunk 1 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp:42]: fixed everywhere.

File: velox/connectors/hive/storage_adapters/gcs/tests/GcsFileSystemTest.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+#include "gtest/gtest.h"
+
+namespace facebook::velox::filesystems {
+namespace {
+
+class GcsFileSystemTest : public testing::Test {
+ public:
+  void SetUp() {
+    testbench_ = std::make_shared<GcsEmulator>();
```
Comment (pedroerp) [inline velox/connectors/hive/storage_adapters/gcs/tests/GcsFileSystemTest.cpp:32]: nit: could also call this variable something else (`emulator_`)?

**Review by pedroerp — APPROVED** [review]
Thanks @majetideepak

**Pull request closed without merging**